### PR TITLE
Update UFS to develop as of 2023 Apr 17

### LINKED ISSUE #1279: Improve diag_table_cpl
**Description**
The coupled model related diag_table is writing out static information regarding ocean geometry in multiple files at multiple frequencies.

E.g. 
```
# static fields
"ocean_model", "geolon",      "geolon",      "ocn%4yr%2mo%2dy%2hr", "all", .false., "none", 2
"ocean_model", "geolat",      "geolat",      "ocn%4yr%2mo%2dy%2hr", "all", .false., "none", 2
"ocean_model", "geolon_c",    "geolon_c",    "ocn%4yr%2mo%2dy%2hr", "all", .false., "none", 2
"ocean_model", "geolat_c",    "geolat_c",    "ocn%4yr%2mo%2dy%2hr", "all", .false., "none", 2
"ocean_model", "geolon_u",    "geolon_u",    "ocn%4yr%2mo%2dy%2hr", "all", .false., "none", 2
"ocean_model", "geolat_u",    "geolat_u",    "ocn%4yr%2mo%2dy%2hr", "all", .false., "none", 2
"ocean_model", "geolon_v",    "geolon_v",    "ocn%4yr%2mo%2dy%2hr", "all", .false., "none", 2
"ocean_model", "geolat_v",    "geolat_v",    "ocn%4yr%2mo%2dy%2hr", "all", .false., "none", 2

# save daily SST
 "ocean_model", "geolon",      "geolon",      "ocn_daily%4yr%2mo%2dy", "all", .false., "none", 2
 "ocean_model", "geolat",      "geolat",      "ocn_daily%4yr%2mo%2dy", "all", .false., "none", 2
 "ocean_model", "geolon_c",    "geolon_c",    "ocn_daily%4yr%2mo%2dy", "all", .false., "none", 2
 "ocean_model", "geolat_c",    "geolat_c",    "ocn_daily%4yr%2mo%2dy", "all", .false., "none", 2
 "ocean_model", "geolon_u",    "geolon_u",    "ocn_daily%4yr%2mo%2dy", "all", .false., "none", 2
 "ocean_model", "geolat_u",    "geolat_u",    "ocn_daily%4yr%2mo%2dy", "all", .false., "none", 2
 "ocean_model", "geolon_v",    "geolon_v",    "ocn_daily%4yr%2mo%2dy", "all", .false., "none", 2
 "ocean_model", "geolat_v",    "geolat_v",    "ocn_daily%4yr%2mo%2dy", "all", .false., "none", 2
```

The `static fields` are written in the `ocn%4yr%2mo%2dy%2hr` file, while the daily fields are written in `ocn_daily%4yr%2mo%2dy`.

This data does not change run-to-run for a given configuration.

These fields are likely small, however, results in disk usage.  This will be exasperated with an 80 member ensemble, run 4x daily, writing output hourly through the DA window.

Determine the use of these fields in post model integration use and determine if this data can be saved as a static file along side other fix files for the model.  

Investigate if these fields can be written out to `ocean_geometry.nc` as that is written out just once.

**Requirements**
- A justification how this data is used and why it should be saved at such a high frequency.

**Acceptance Criteria (Definition of Done)**
- Simplify `diag_table_cpl` and harmonize with `diag_table` 
- Eliminate repeated duplicative static information.

**Dependencies**
None.

### LINKED ISSUE #1445: Add new capability of writing restart files on the write grid comp
**Description**
The new capability of writing restart files on the write grid components is available in ufs WM ([PR#1633](https://github.com/ufs-community/ufs-weather-model/pull/1633)), committed on 3/28/23). The new feature needs to be integrated in the global_workflow to be used by GFSv17/GEFSv13.

To use the feature, the following changes are required: 
- Add following variable in model_configure: 
quilting_restart:        @[QUILTING_RESTART]
- QUILTING_RESTART:  
1) when .false., the restart files will be written out by FMS during the forecast time (current way)
2) when .true., the restart files will be written out through write grid component with asynchronized IO.

**Requirements**

**Acceptance Criteria (Definition of Done)**

**(Optional): Suggest A Solution**

### LINKED ISSUE #1499: Update ufs-weather-model (to EP4 tag)
**Description**
Update ufs-weather-model to EP4 tag (which is only a few commits behind the top of develop of ufs-weather-model and will be used by GEFS group for EP4).  

The updates of note for this include: 
- Updates to unified precip output types (will require diag_table update) 
- Bug fix for convection
- Bug fix for gravity wave drag that causes stochastic physics runs to crash (this is needed for ensemble runs)  

**Requirements**
- Update model tag

**Acceptance Criteria (Definition of Done)**
Using the new tag of ufs-weather-model in develop branch. 

**(Optional): Suggest A Solution**
Update ufs-weather-model to EP4 tag and make sure other files are up-to-date with new model version.

### PR DESCRIPTION
**Pull request by WalterKolczynski-NOAA**
**Description**
Updates the UFS model hash to the version as of 2023 Apr 17.

Some associated changes accompany this update
- Restart filenames for MOM6 (ufs-community/ufs-weather-model#1599)
- Remove store coriolis setting from MOM6 namelist (ufs-community/ufs-weather-model#1599)
- Change in atm 'log' file names (ufs-community/ufs-weather-model#1704)
- Additions to diag_table for frozen species (ufs-community/ufs-weather-model#1529)
- Restart quilting (ufs-community/ufs-weather-model#1633)
- Update to post itag (ufs-community/ufs-weather-model#1690)

The switch to restart quilting adds an additional constraint on the size of write groups to be divisible by the number of tiles, so all were increased to the next multiple of 6.

In the process of updating the diag tables, unused tables were removed.

Closes #1279 as moot
Closes #1445
Closes #1499
Partially addresses #1277

**Type of change**
- [x] New feature (non-breaking change which adds functionality)

**How Has This Been Tested?**
- [x] Cycled test on Orion
- [x] P8 test on Orion
  
**Checklist**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] Any dependent changes have been merged and published

### CODE REVIEW

**Comment by JessicaMeixner-NOAA** [comment]
Thanks for the PR @WalterKolczynski-NOAA ! I'll review in the morning!

File: parm/parm_fv3diag/diag_table_cpl
Comment (aerorahul) [inline parm/parm_fv3diag/diag_table_cpl:1]: I don't think this `diag_table_cpl` is used anymore.
I suggest we delete it as well.

File: parm/parm_fv3diag/diag_table_da [hunk 1]
```diff
@@ -34,6 +34,13 @@
 "gfs_dyn",     "hs",          "hgtsfc",       "fv3_history",    "all",  .false.,  "none",  2
 
 "gfs_phys",    "cldfra",        "cldfra",        "fv3_history2d",  "all",  .false.,  "none",  2
+"gfs_phys",    "frzr",          "frzr",          "fv3_history2d",  "all",  .false.,  "none",  2
+"gfs_phys",    "frzrb",         "frzrb",         "fv3_history2d",  "all",  .false.,  "none",  2
+"gfs_phys",    "frozr",         "frozr",         "fv3_history2d",  "all",  .false.,  "none",  2
+"gfs_phys",    "frozrb",        "frozrb",        "fv3_history2d",  "all",  .false.,  "none",  2
+"gfs_phys",    "tsnowp",        "tsnowp",        "fv3_history2d",  "all",  .false.,  "none",  2
+"gfs_phys",    "tsnowpb",       "tsnowpb",       "fv3_history2d",  "all",  .false.,  "none",  2
+"gfs_phys",    "rhonewsn",      "rhonewsn",      "fv3_history2d",  "all",  .false.,  "none",  2
```
Comment (aerorahul) [inline parm/parm_fv3diag/diag_table_da]: Does the GSI read/need these? 
@CatherineThomas-NOAA @emilyhcliu 
I suggest we don't increase the volume of our output unless this is information needed by the DA executables.

File: parm/config/config.resources [hunk 2]
```diff
@@ -522,8 +522,7 @@ elif [[ "${step}" = "fcst" || "${step}" = "efcs" ]]; then
         # PETS for quilting
         if [[ "${QUILTING:-}" = ".true." ]]; then
           (( QUILTPETS = ntasks_quilt * nthreads_fv3 ))
-          (( WRTTASK_PER_GROUP = WRTTASK_PER_GROUP * nthreads_fv3 ))  # when threads are used, WRTTASKS_PER_GROUP = INCOMING_WRTTASKS_PER_GROUP * threads
-          # model_configure should be updated to reflect the number of threads used by calling this variable write_pets_per_group
+          (( WRTTASK_PER_GROUP = WRTTASK_PER_GROUP ))
```
Comment (aerorahul) [inline parm/config/config.resources]: This is incorrect.
The `WRTTASKS_PER_GROUP` need to account the threads the atmosphere is running with.

**Review by aerorahul — DISMISSED** [review]
Please see the inline comments.

File: parm/config/config.resources (hunk 2 above)
Comment (WalterKolczynski-NOAA) [reply parm/config/config.resources]: Not anymore, apparently. I only made this change after the forecast failed due to a PE mismatch. This line is rather superfluous now though, I should've just deleted it.

File: parm/parm_fv3diag/diag_table_da (hunk 1 above)
Comment (WalterKolczynski-NOAA) [reply parm/parm_fv3diag/diag_table_da]: I wasn't sure myself, but forgot I wanted to ask DA about them when I opened the PR. Thanks Rahul.

File: parm/config/config.resources (hunk 2 above)
Comment (WalterKolczynski-NOAA) [reply parm/config/config.resources]: Maybe @junwang-noaa can clarify?

File: parm/parm_fv3diag/diag_table_cpl
Comment (WalterKolczynski-NOAA) [reply parm/parm_fv3diag/diag_table_cpl:1]: Done

**Comment by WalterKolczynski-NOAA** [comment]
The diag_table additions get rid of several of the undefined variables @TerrenceMcGuinness-NOAA discovered in the grib2 files. Only `REFC` and `REFD` remain (simulated radar). @WenMeng-NOAA, is it possible we are missing other diag_table entries for these two?

**Comment by WalterKolczynski-NOAA** [comment]
> The diag_table additions get rid of several of the undefined variables @TerrenceMcGuinness-NOAA discovered in the grib2 files. Only `REFC` and `REFD` remain (simulated radar). @WenMeng-NOAA, is it possible we are missing other diag_table entries for these two?

Created a new issue for this: #1513

**Comment by WenMeng-NOAA** [comment]
> > The diag_table additions get rid of several of the undefined variables @TerrenceMcGuinness-NOAA discovered in the grib2 files. Only `REFC` and `REFD` remain (simulated radar). @WenMeng-NOAA, is it possible we are missing other diag_table entries for these two?
> 
> Created a new issue for this: #1513

The UPP read 3D "refl_10cm" from model output gfs*sfcf*. @WalterKolczynski-NOAA

File: parm/parm_fv3diag/diag_table_da (hunk 1 above)
Comment (CatherineThomas-NOAA) [reply parm/parm_fv3diag/diag_table_da]: No, the GSI only reads the 3D hydrometeors.  

@jiaruidong2017 @CoryMartin-NOAA - Does the snow DA use these 2D variables at all?

Comment (CoryMartin-NOAA) [reply parm/parm_fv3diag/diag_table_da]: snow DA is currently using FMS restarts. These are the sfcfNNN history files, right? If so, snow DA is unaffected by this table.

Comment (aerorahul) [reply parm/parm_fv3diag/diag_table_da]: Thanks @CoryMartin-NOAA @CatherineThomas-NOAA 
We will remove adding these fields for now.
Please open an issue if you wish to add these back in to the surface Gaussian files for the GDAS cycle.

Comment (WalterKolczynski-NOAA) [reply parm/parm_fv3diag/diag_table_da]: Removed. Thanks everyone.

File: parm/config/config.resources (hunk 2 above)
Comment (junwang-noaa) [reply parm/config/config.resources]: What I see here is that the WRTTASKS_PER_GROUP is the multiple of the MPI write tasks per group * nthreads_fv3. In other words, inside the write grid comp, the number of MPI tasks is actually WRTTASKS_PER_GROUP/nthreads_fv3, which I think is the original WRTTASKS_PER_GROUP.

**Review by WenMeng-NOAA — APPROVED** [review]
The updates for the post process part looks good for me.

File: parm/config/config.resources (hunk 2 above)
Comment (aerorahul) [reply parm/config/config.resources]: I suggest just delete this line.  It is superfluous as you note.
There is some major confusion regarding threading in the write task calculation.
```suggestion
```

**Comment by JessicaMeixner-NOAA** [comment]
To get my C768 HR1 like forecast job to run, I needed to double the number of WRTTASK_PER_GROUP_PER_THREAD_PER_TILE 
(there's perhaps a smaller amount to run on hera, plus we still need netcdf_parallel -> netcdf unless that error was solved and I just changed this debugging the other).  

Then I also needed this change: 
```
diff --git a/parm/config/config.resources b/parm/config/config.resources
index cf010f58..1e5b7479 100644
--- a/parm/config/config.resources
+++ b/parm/config/config.resources
@@ -530,7 +530,7 @@ elif [[ "${step}" = "fcst" || "${step}" = "efcs" ]]; then
         # PETS for quilting
         if [[ "${QUILTING:-}" = ".true." ]]; then
           (( QUILTPETS = ntasks_quilt * nthreads_fv3 ))
-          (( WRTTASK_PER_GROUP = WRTTASK_PER_GROUP_PER_THREAD * nthreads_fv3 ))
+          (( WRTTASK_PER_GROUP = WRTTASK_PER_GROUP_PER_THREAD ))
           export WRTTASK_PER_GROUP
         else
           QUILTPETS=0
```


This is not necessarily related to this PR, but if you change the number of threads for ice, we need this (let me know I can make a separate issue/PR for this if desired):

```
diff --git a/ush/parsing_namelists_CICE.sh b/ush/parsing_namelists_CICE.sh
index cf5b0fbf..6fc8b143 100755
--- a/ush/parsing_namelists_CICE.sh
+++ b/ush/parsing_namelists_CICE.sh
@@ -25,8 +25,8 @@ fi
 # Get correct MPI options for NPROC and grid
 local processor_shape=${cice6_processor_shape:-'slenderX2'}
 local shape=${processor_shape#${processor_shape%?}}
-local NPX=$(( ICEPETS / shape )) #number of processors in x direction
-local NPY=$(( ICEPETS / NPX ))   #number of processors in y direction
+local NPX=$(( ntasks_cice6 / shape )) #number of processors in x direction
+local NPY=$(( ntasks_cice6 / NPX ))   #number of processors in y direction
 if (( $(( NX_GLB % NPX )) == 0 )); then
   local block_size_x=$(( NX_GLB / NPX ))
 else
@@ -201,7 +201,7 @@ cat > ice_in <<eof
 /
 
 &domain_nml
-   nprocs = ${ICEPETS}
+   nprocs            = ${ntasks_cice6}
    nx_global         = ${NX_GLB}
```

**Comment by JessicaMeixner-NOAA** [comment]
This isn't from this PR, but I ran into a bug testing this PR in which line [182 of exglobal_archive.sh](https://github.com/NOAA-EMC/global-workflow/blob/develop/scripts/exglobal_archive.sh#L182) should be updated from: 
if [ "${DO_WAVE}" = "YES" ] && [ "${WAVE_RUN}" != "gdas" ]; then
to 
if [ "${DO_WAVE}" = "YES" ] ; then

as we've already checked $RUN is gfs before, so we shouldn't need that check here and WAVE_RUN is not defined and causes an error in the archive.  Testing this fix right now.

**Review by JessicaMeixner-NOAA — CHANGES_REQUESTED** [review]
A few bug fixes are needed, but thank you so much for making this PR @WalterKolczynski-NOAA

**Comment by JessicaMeixner-NOAA** [comment]
> This isn't from this PR, but I ran into a bug testing this PR in which line [182 of exglobal_archive.sh](https://github.com/NOAA-EMC/global-workflow/blob/develop/scripts/exglobal_archive.sh#L182) should be updated from: if [ "${DO_WAVE}" = "YES" ] && [ "${WAVE_RUN}" != "gdas" ]; then to if [ "${DO_WAVE}" = "YES" ] ; then
> 
> as we've already checked $RUN is gfs before, so we shouldn't need that check here and WAVE_RUN is not defined and causes an error in the archive. Testing this fix right now.

This fix works.  Please let me know if you'd like this fix in a new PR, happy to submit it separately if that's preferred as it's not really related to this PR.

**Comment by JessicaMeixner-NOAA** [comment]
@WalterKolczynski-NOAA Bing let me know that all the new output from the diag_table updates look good.  Please let me know if you want me to push any of these bug fixes that are not really related to this PR but found during it's testing to a new PR or to your branch for this PR, etc.

**Comment by WalterKolczynski-NOAA** [comment]
> @WalterKolczynski-NOAA Bing let me know that all the new output from the diag_table updates look good. Please let me know if you want me to push any of these bug fixes that are not really related to this PR but found during it's testing to a new PR or to your branch for this PR, etc.

Just open new issues and I'll handle them on Monday (off today).

**Comment by JessicaMeixner-NOAA** [comment]
Thanks @WalterKolczynski-NOAA I created #1548 and #1549.  Line 533 in config.resources in this PR still needs to be fixed before merging I believe.

**Review by JessicaMeixner-NOAA — APPROVED** [review]
Thanks for fixing Line 533.

**Review by aerorahul — APPROVED** [review]
I don't know what the updates in the ufs-weather-model are that the `model_configure` needs.  But since these edits make it work, I am ok with it.

File: parm/config/config.ufs [hunk 3]
```diff
@@ -197,23 +197,28 @@ case "${fv3_res}" in
         export nthreads_fv3_gfs=4
         export cdmbgwd="4.0,0.05,1.0,1.0"  # mountain blocking, ogwd, cgwd, cgwd src scaling
         export WRITE_GROUP=4
-        export WRTTASK_PER_GROUP=64  # TODO: refine these numbers when a case is available
+        export WRTTASK_PER_GROUP_PER_THREAD_PER_TILE=10  # TODO: refine these numbers when a case is available
         export WRITE_GROUP_GFS=4
-        export WRTTASK_PER_GROUP_GFS=64  # TODO: refine these numbers when a case is available
+        export WRTTASK_PER_GROUP_PER_THREAD_PER_TILE_GFS=10  # TODO: refine these numbers when a case is available
```
Comment (junwang-noaa) [inline parm/config/config.ufs:202]: Can we change this to a multiple of 6, e,g, 12? It gives developers option to try writing restart files on write grid comp?

Comment (aerorahul) [reply parm/config/config.ufs:202]: As seen from the variable name, it is per tile. So, it gets multiplied by 6 before ending up in `model_configure`.

Comment (WalterKolczynski-NOAA) [reply parm/config/config.ufs:202]: @junwang-noaa The meaning of the variable has already been changed to be per tile (we multiply by 6 later), so it is impossible for the end value to not be divisible by 6.

Comment (junwang-noaa) [reply parm/config/config.ufs:202]: Thanks for the explanation.

**Review by junwang-noaa — APPROVED** [review]

**Pull request merged**